Support for autoscale for HDInsight clusters

kosinsky opened the pull request:
Adds autoscale (capacity and scheduled) support for HDInsight clusters. Following resources were updated :

* azurerm_hdinsight_hadoop_cluster
* azurerm_hdinsight_hbase_cluster (scheduled only)
* azurerm_hdinsight_interactive_query_cluster
* azurerm_hdinsight_spark_cluster

Others weren't modified, because:
* Spark doesn't support autoscale
* Storm and ML/RServer cluster types are deprecated and will retired at the end of 2020

Capacity based autoscale:
```hcl
      autoscale {
        capacity {
          min_instance_count = 2
          max_instance_count = 3
        }
      }
```

Scheduled autoscale:
```hcl
      autoscale {
        recurrence {
          timezone  = "Pacific Standard Time"          
          schedule {
            days      = ["Monday"]
            time      = "10:00"
            target_instance_count = 5
          }
          schedule {
            days      = ["Saturday", "Sunday"]
            time      = "10:00"
            target_instance_count = 3
          }
        }
      }
```

Tests:
```bash
make testacc TEST='./azurerm/internal/services/hdinsight/tests/' TESTARGS='-run=autoscale -parallel=3' TESTTIMEOUT='300m'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test ./azurerm/internal/services/hdinsight/tests/ -v -run=autoscale -parallel=3 -timeout 300m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccAzureRMHDInsightHadoopCluster_autoscale
=== PAUSE TestAccAzureRMHDInsightHadoopCluster_autoscale
=== RUN   TestAccAzureRMHDInsightHBaseCluster_autoscale
=== PAUSE TestAccAzureRMHDInsightHBaseCluster_autoscale
=== RUN   TestAccAzureRMHDInsightInteractiveQueryCluster_autoscale
=== PAUSE TestAccAzureRMHDInsightInteractiveQueryCluster_autoscale
=== RUN   TestAccAzureRMHDInsightSparkCluster_autoscale
=== PAUSE TestAccAzureRMHDInsightSparkCluster_autoscale
=== CONT  TestAccAzureRMHDInsightHadoopCluster_autoscale
=== CONT  TestAccAzureRMHDInsightSparkCluster_autoscale
=== CONT  TestAccAzureRMHDInsightInteractiveQueryCluster_autoscale
--- PASS: TestAccAzureRMHDInsightInteractiveQueryCluster_autoscale (1480.93s)
=== CONT  TestAccAzureRMHDInsightHBaseCluster_autoscale
--- PASS: TestAccAzureRMHDInsightHadoopCluster_autoscale (1853.24s)
--- PASS: TestAccAzureRMHDInsightSparkCluster_autoscale (2087.36s)
--- PASS: TestAccAzureRMHDInsightHBaseCluster_autoscale (1318.89s)
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/hdinsight/tests     2800.965s
```

Fixes #6470

jackofallops commented on azurerm/internal/services/hdinsight/common_hdinsight.go:
```diff
@@ -47,18 +47,37 @@ func hdinsightClusterUpdate(clusterKind string, readFunc schema.ReadFunc) schema
-			if err = future.WaitForCompletionRef(ctx, client.Client); err != nil {
-				return fmt.Errorf("Error waiting for the HDInsight %q Cluster %q (Resource Group %q) to finish resizing: %+v", clusterKind, name, resourceGroup, err)
+			if d.HasChange("roles.0.worker_node.0.autoscale") {
+				autoscale := azure.ExpandHDInsightNodeAutoScaleDefinition(workerNode["autoscale"].([]interface{}))
```
This will need adjusting after the rebase
```suggestion
				autoscale := ExpandHDInsightNodeAutoScaleDefinition(workerNode["autoscale"].([]interface{}))
```

jackofallops commented on azurerm/internal/services/hdinsight/tests/hdinsight_hadoop_cluster_resource_test.go:
```diff
@@ -1163,6 +1219,8 @@ func testAccAzureRMHDInsightHadoopCluster_template(data acceptance.TestData) str
 	return fmt.Sprintf(`
 provider "azurerm" {
   features {}
+  # Due to bug in HDInsight we can't reuse client with the same x-ms-correlation-request-id for multiple updates
+  disable_correlation_request_id = true
```
Is there a public GH issue tracking this for resolution that we can leave in the comments here?

kosinsky replied:
I'm not aware about GH repo with public HDI issue. I got that information from Microsoft support and then Product group when I hit the issue. It's more like limitation than a bug. AFAIK, it's by design.

tombuildsstuff replied:
rather than doing this within the test suite, we should be able to update the [client](https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/internal/services/hdinsight/client/client.go#L15-L34) to not send this:

```
func NewClient(o *common.ClientOptions) *Client {
	// due to a bug in the HDInsight API we can't reuse client with the same x-ms-correlation-request-id for multiple updates
	opts := *o
	opts.DisableCorrelationRequestID = true

	ApplicationsClient := hdinsight.NewApplicationsClientWithBaseURI(opts.ResourceManagerEndpoint, opts.SubscriptionId)
	opts.ConfigureClient(&ApplicationsClient.Client, opts.ResourceManagerAuthorizer)

	ClustersClient := hdinsight.NewClustersClientWithBaseURI(opts.ResourceManagerEndpoint, opts.SubscriptionId)
	opts.ConfigureClient(&ClustersClient.Client, opts.ResourceManagerAuthorizer)

	ConfigurationsClient := hdinsight.NewConfigurationsClientWithBaseURI(opts.ResourceManagerEndpoint, opts.SubscriptionId)
	opts.ConfigureClient(&ConfigurationsClient.Client, opts.ResourceManagerAuthorizer)

	ExtensionsClient := hdinsight.NewExtensionsClientWithBaseURI(opts.ResourceManagerEndpoint, opts.SubscriptionId)
	opts.ConfigureClient(&ExtensionsClient.Client, opts.ResourceManagerAuthorizer)

	return &Client{
		ApplicationsClient:   &ApplicationsClient,
		ClustersClient:       &ClustersClient,
		ConfigurationsClient: &ConfigurationsClient,
		ExtensionsClient:     &ExtensionsClient,
	}
}
```

kosinsky replied:
Done.

jackofallops reviewed (changes requested):
Hi @kosinsky 

Thanks for this PR. It's looking good.  We've done some refactor/tidying work that have moved some of the helper code around, so this will need a rebase (it should do so cleanly I think).  There's one change I think will need to be made after that (commented below).  Once that's done I'll run the CI tests and I think this should be fine as is.  There's an additional question on the bug you mention which would be good to have an answer on.

Thanks again.

kosinsky commented:
Merge and post merge refactoring reactions are done. Please, take a look

tombuildsstuff commented on website/docs/r/hdinsight_hbase_cluster.html.markdown:
```diff
@@ -291,6 +293,32 @@ A `monitor` block supports the following:
+
+* `recurrence` - (Optional) A `recurrence` block as defined below.
+
+-> **NOTE:** Capacity based autoscaling isn't supported to HBase clusters.
```
we can remove this and conditionally make the `recurrence` block required here?

kosinsky replied:
I made option required. However, purpose of the Note is to emphasis on a difference between HBase and other cluster types.

tombuildsstuff commented on website/docs/r/hdinsight_hbase_cluster.html.markdown:
```diff
@@ -291,6 +293,32 @@ A `monitor` block supports the following:
+
+* `timezone` - (Required) The time zone for the autoscale schedule times.
+
+* `schedule` - A list of `schedule` blocks as defined below.
```
```suggestion
* `schedule` - (Required) A list of `schedule` blocks as defined below.
```

tombuildsstuff commented on website/docs/r/hdinsight_interactive_query_cluster.html.markdown:
```diff
@@ -294,6 +296,42 @@ A `monitor` block supports the following:
+
+* `recurrence` - (Optional) A `recurrence` block as defined below.
+
+-> **NOTE:** Either a `capacity` or `recurrence` must be specified - but not both.
```
```suggestion
-> **NOTE:** Either a `capacity` or `recurrence` block must be specified - but not both.
```

tombuildsstuff commented on website/docs/r/hdinsight_interactive_query_cluster.html.markdown:
```diff
@@ -294,6 +296,42 @@ A `monitor` block supports the following:
+
+* `min_instance_count` - (Required) The minimum number of worker nodes to autoscale to based on the cluster's activity.
+
+* `max_instance_count` - (Required) The maximum number of worker nodes to autoscale to based on the cluster's activity.
```
could we sort these alphabetically?

tombuildsstuff commented on website/docs/r/hdinsight_interactive_query_cluster.html.markdown:
```diff
@@ -294,6 +296,42 @@ A `monitor` block supports the following:
+
+* `timezone` - (Required) The time zone for the autoscale schedule times.
+
+* `schedule` - A list of `schedule` blocks as defined below.
```
```suggestion
* `schedule` - (Required) A list of `schedule` blocks as defined below.
```

tombuildsstuff commented on website/docs/r/hdinsight_spark_cluster.html.markdown:
```diff
@@ -291,6 +293,42 @@ A `monitor` block supports the following:
+
+* `recurrence` - (Optional) A `recurrence` block as defined below.
+
+-> **NOTE:** Either a `capacity` or `recurrence` must be specified - but not both.
```
```suggestion
-> **NOTE:** Either a `capacity` or `recurrence` block must be specified - but not both.
```

tombuildsstuff commented on website/docs/r/hdinsight_spark_cluster.html.markdown:
```diff
@@ -291,6 +293,42 @@ A `monitor` block supports the following:
+
+* `min_instance_count` - (Required) The minimum number of worker nodes to autoscale to based on the cluster's activity.
+
+* `max_instance_count` - (Required) The maximum number of worker nodes to autoscale to based on the cluster's activity.
```
could we sort these alphabetically?

tombuildsstuff commented on website/docs/r/hdinsight_spark_cluster.html.markdown:
```diff
@@ -291,6 +293,42 @@ A `monitor` block supports the following:
+
+* `timezone` - (Required) The time zone for the autoscale schedule times.
+
+* `schedule` - A list of `schedule` blocks as defined below.
```
```suggestion
* `schedule` - (Required) A list of `schedule` blocks as defined below.
```

kosinsky replied:
Schedule Required

tombuildsstuff reviewed (changes requested):
hey @kosinsky 

Thanks for pushing those changes :)

I've taken a look through and left some comments inline but this is looking good - if we can fix up the comments then this should otherwise be good to merge 👍 

Thanks!

kosinsky commented:
@tombuildsstuff @jackofallops I addressed your feedback. Sorry for delay.

sseppola commented:
@kosinsky thank you for all this work! It's much appreciated.

In case this was drowned in other notifications and forgotten about I wanted to leave a comment in hope it attracts some attention to a nearly finished PR :)

kosinsky commented:
@tombuildsstuff @jackofallops Could you take one more look? I addressed your feedback. Please, let me know if I missed something or you have additional one.

mbfrahry reviewed (changes requested):
Hey @kosinsky! I got a chance to look over this and it looks good! The only issue is that it's quite a bit behind...through no fault of your own. Once it's back up to speed, we can get it merged quickly.

katbyte commented:
@kosinsky - going to close this for #11547 - Thanks for getting it started!
